### PR TITLE
 For #5694 & #6054: Allows users to change toolbar position

### PR DESCRIPTION
**Pull request by sblatz**
---

### Pull Request checklist

- [x] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Screenshots**: This PR includes screenshots or GIFs of the changes made or an explanation of why it does not
- [x] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features

### After merge
- [ ] **Milestone**: Make sure issues finished by this pull request are added to the [milestone](https://github.com/mozilla-mobile/fenix/milestones) of the version currently in development.

### To download an APK when reviewing a PR:
1. click on Show All Checks,
2. click Details next to "Taskcluster (pull_request)" after it appears and then finishes with a green checkmark,
3. click on the "Fenix - assemble" task, then click "Run Artifacts".
4. the APK links should be on the left side of the screen, named for each CPU architecture

### CODE REVIEW

**Comment by sblatz** [comment]
# Request for data collection review form

**All questions are mandatory. You must receive review from a data steward peer on your responses to these questions before shipping new data collection.**

1) **What questions will you answer with this data?**
- How often users choose a top toolbar. 
- How many users are currently using top vs bottom toolbar (core metric ping)

2) **Why does Mozilla need to answer these questions?  Are there benefits for users? Do we need this information to address product or business requirements?**

- If more than 20% of users choose top nav bar (without surfacing this option during onboarding) then we will investigate (through user research or otherwise) whether we need to consider changing the default.

3) **What alternative methods did you consider to answer these questions? Why were they not sufficient?**

* N/A (These are baseline metrics)

4) **Can current instrumentation answer these questions?**
* No, this is a new feature and there is no telemetry surrounding which selection users make for their toolbar position.

5) **List all proposed measurements and indicate the category of data collection for each measurement, using the Firefox [data c](https://wiki.mozilla.org/Firefox/Data_Collection)[ollection ](https://wiki.mozilla.org/Firefox/Data_Collection)[categories](https://wiki.mozilla.org/Firefox/Data_Collection) on the found on the Mozilla wiki.**

* All data is Category 2.

6) **How long will this data be collected?**

* Until 03/01/2020

7) **What populations will you measure?**

* All release, beta, and nightly users with telemetry enabled.

8) **Please provide a general description of how you will analyze this data.**

* Glean / Amplitude

9) **Where do you intend to share the results of your analysis?**

* Only on Glean, Amplitude and with mobile teams.

File: app/src/main/java/org/mozilla/fenix/components/toolbar/EngineViewTopBehavior.kt [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+        return super.layoutDependsOn(parent, child, dependency)
+    }
+
+    /**
+     * Apply vertical clipping to [EngineView]. This requires [EngineViewTopehavior] to be set
+     * in/on the [EngineView] or its parent. Must be a direct descending child of [CoordinatorLayout].
+     */
+    override fun onDependentViewChanged(parent: CoordinatorLayout, child: View, dependency: View): Boolean {
+        val engineView = child.findViewInHierarchy { it is EngineView } as EngineView?
+        engineView?.setVerticalClipping(dependency.height - dependency.translationY.toInt())
```
Comment (hiikezoe) [inline app/src/main/java/org/mozilla/fenix/components/toolbar/EngineViewTopBehavior.kt]: This looks pretty odd to me.  This `setVerticalClipping` API ends up [calling `setFixedBottomOffset` in GeckoView](https://searchfox.org/mozilla-central/rev/652014ca1183c56bc5f04daf01af180d4e50a91c/mobile/android/geckoview/src/main/java/org/mozilla/geckoview/GeckoDisplay.java#140-144), the function does just set a margin for position:fixed elements at bottom of viewport. As far as I can tell, we need at least one new API make the top toolbar dynamic properly, the API would be a function to set an offset of the root viewport (i.e. not only for position:fixed elements).

CCing @pocmo and @snorp.

File: app/src/main/java/org/mozilla/fenix/browser/BrowserFragment.kt [hunk 2]
```diff
@@ -128,6 +130,17 @@ class BrowserFragment : BaseBrowserFragment(), BackHandler {
...
+    }
+
+    private fun updateEngineBottomMargin() {
+        val browserEngine = swipeRefresh.layoutParams as CoordinatorLayout.LayoutParams
+
+        browserEngine.bottomMargin = if (requireContext().settings().shouldUseBottomToolbar) {
+            requireContext().dimen(R.dimen.browser_toolbar_height)
+        } else {
+            0
+        }
```
Comment (pocmo) [inline app/src/main/java/org/mozilla/fenix/browser/BrowserFragment.kt]: Is this margin still needed?

File: app/src/main/java/org/mozilla/fenix/components/toolbar/EngineViewTopBehavior.kt (hunk 1 above)
Comment (pocmo) [reply app/src/main/java/org/mozilla/fenix/components/toolbar/EngineViewTopBehavior.kt]: Yes, for the top toolbar this should not be called at all. And I think @sblatz decided to not use this behavior implementation for the top toolbar, right? Afaik this was never really used in Fenix.

File: app/src/main/java/org/mozilla/fenix/browser/BrowserFragment.kt (hunk 2 above)
Comment (sblatz) [reply app/src/main/java/org/mozilla/fenix/browser/BrowserFragment.kt]: @pocmo Yes, if I don't do this, the GV will be covered by the bottom toolbar.

**Comment by ekager** [comment]
Could we make sure we're dealing with the snackbar anchoring correctly in browser fragments? 

Maybe add a new function that gets the anchor by checking the status like
`if (settings().shouldUseBottomToolbar) (toolbar.view) else null`

File: app/src/main/java/org/mozilla/fenix/utils/Settings.kt [hunk 3]
```diff
@@ -198,6 +199,12 @@ class Settings private constructor(
         true
     )
 
+    val shouldUseFixedToolbar: Boolean
+        get() {
+            val accessibilityManager = appContext.getSystemService(Context.ACCESSIBILITY_SERVICE) as? AccessibilityManager
```
Comment (Baron-Severin) [inline app/src/main/java/org/mozilla/fenix/utils/Settings.kt]: What does fixed toolbar mean, and how does it relate to accessibility?  Would you add a short (1 or 2 lines) comment here explaining?

File: app/src/main/java/org/mozilla/fenix/search/toolbar/ToolbarView.kt [hunk 4]
```diff
@@ -58,9 +59,21 @@ class ToolbarView(
...
-        .inflate(R.layout.component_search, container, true)
-        .findViewById(R.id.toolbar)
+    val view: BrowserToolbar = if (container.context.settings().shouldUseBottomToolbar) {
+        LayoutInflater.from(container.context)
+            .inflate(R.layout.component_bottom_browser_toolbar, container, true)
+            .findViewById(R.id.toolbar_bottom)
+    } else {
+        if (container.context.settings().shouldUseFixedToolbar) {
+            LayoutInflater.from(container.context)
+                .inflate(R.layout.component_browser_top_toolbar, container, true)
```
Comment (Baron-Severin) [inline app/src/main/java/org/mozilla/fenix/search/toolbar/ToolbarView.kt]: I think this needs to += `_fixed`.  If not, please write a comment explaining why.

File: app/src/main/res/layout/component_browser_bottom_toolbar.xml [hunk 5]
```diff
@@ -0,0 +1,25 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (Baron-Severin) [inline app/src/main/res/layout/component_browser_bottom_toolbar.xml]: Nit: license

File: app/src/main/res/layout/component_browser_top_toolbar_fixed.xml [hunk 6]
```diff
@@ -0,0 +1,25 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (Baron-Severin) [inline app/src/main/res/layout/component_browser_top_toolbar_fixed.xml]: Nit: license

File: app/src/main/res/layout/component_browser_top_toolbar_fixed.xml [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+        android:clickable="true"
+        android:focusable="true"
+        android:focusableInTouchMode="true"
+        app:browserToolbarClearColor="?primaryText"
+        app:browserToolbarInsecureColor="?primaryText"
+        app:browserToolbarMenuColor="?primaryText"
+        app:browserToolbarProgressBarGravity="top"
+        app:browserToolbarSecureColor="?primaryText"
+        app:browserToolbarTrackingProtectionAndSecurityIndicatorSeparatorColor="?toolbarDivider" />
+</com.google.android.material.appbar.AppBarLayout>
```
Comment (Baron-Severin) [inline app/src/main/res/layout/component_browser_top_toolbar_fixed.xml]: Is the only difference between these two layouts `app:layout_scrollFlags`?  Normally I'm not a fan of setting layout values programmatically, but it seems more confusing here to split these into different files.  Would it be possible to consolidate these and just null out the scroll flags when fixed mode is on?

File: app/metrics.yaml [hunk 8]
```diff
@@ -17,7 +17,7 @@ events:
       source:
         description: "The method used to open Fenix. Possible values are: `app_icon`, `custom_tab` or `link`"
     bugs:
-    - https://github.com/mozilla-mobile/fenix/issue/968
+    - https://github.com/mozilla-mobile/fenix/issues/968
```
Comment (Baron-Severin) [inline app/metrics.yaml]: 😬

File: app/src/androidTest/java/org/mozilla/fenix/ui/robots/NavigationToolbarRobot.kt [hunk 9]
```diff
@@ -29,7 +29,7 @@ class NavigationToolbarRobot {
         val mDevice = UiDevice.getInstance(InstrumentationRegistry.getInstrumentation())
```
Comment (Baron-Severin) [inline app/src/androidTest/java/org/mozilla/fenix/ui/robots/NavigationToolbarRobot.kt]: (unrelated to this file)

It felt a little jarring to me when I changed my toolbar setting to top, then went back to the home screen and saw it was still on the bottom.  @apbitner do we have any plans to move the home toolbar as well?

Comment (apbitner) [reply app/src/androidTest/java/org/mozilla/fenix/ui/robots/NavigationToolbarRobot.kt]: Not at the moment. On the home screen we recently moved the search bar from the middle of the screen to the bottom but we can revisit in the future now that we have the dynamic toolbar.

File: app/src/main/java/org/mozilla/fenix/settings/ToolbarSettingsFragment.kt [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+        (activity as AppCompatActivity).title = getString(R.string.preferences_toolbar)
+        (activity as AppCompatActivity).supportActionBar?.show()
+
+        setupClickListeners()
+        setupRadioGroups()
+    }
+
+    private fun setupClickListeners() {
+        val keyToolbarTop = getPreferenceKey(R.string.pref_key_toolbar_top)
+        topPreference = requireNotNull(findPreference(keyToolbarTop))
```
[the remark this answers is not in the document]
Comment (sblatz) [reply app/src/main/java/org/mozilla/fenix/settings/ToolbarSettingsFragment.kt]: I've been told that `requireNotNull` gives us better crash reports in Sentry :)

File: app/src/main/java/org/mozilla/fenix/components/toolbar/BrowserToolbarView.kt [hunk 11]
```diff
@@ -41,19 +40,29 @@ interface BrowserToolbarViewInteractor {
...
+            .findViewById(R.id.toolbar_bottom)
+    } else {
+        if (container.context.settings().shouldUseFixedToolbar) {
+            LayoutInflater.from(container.context)
+                .inflate(R.layout.component_browser_top_toolbar_fixed, container, true)
+                .findViewById(R.id.toolbar_top_fixed)
+        } else {
+            LayoutInflater.from(container.context)
+                .inflate(R.layout.component_browser_bottom_toolbar, container, true)
+                .findViewById(R.id.toolbar_top)
```
[the remark this answers is not in the document]
Comment (sblatz) [reply app/src/main/java/org/mozilla/fenix/components/toolbar/BrowserToolbarView.kt]: This is a much cleaner solution and will lead to far fewer errors with the misuse of the id's. Thanks for this suggestion 😄

File: app/src/main/java/org/mozilla/fenix/search/toolbar/ToolbarView.kt (hunk 4 above)
Comment (sblatz) [reply app/src/main/java/org/mozilla/fenix/search/toolbar/ToolbarView.kt]: 👌 fixed by updating as your other comment mentioned.

File: app/src/main/res/layout/component_browser_top_toolbar_fixed.xml (hunk 7 above)
Comment (sblatz) [reply app/src/main/res/layout/component_browser_top_toolbar_fixed.xml]: Good call, going to set this dynamically :)

**Comment by boek** [comment]
# Data Review Form (to be filled by Data Stewards)

1) Is there or will there be **documentation** that describes the schema for the ultimate data set in a public, complete, and accurate way?
Yes, metrics.yaml and metrics.md

2) Is there a control mechanism that allows the user to turn the data collection on and off?
Yes, under data controls

3) If the request is for permanent data collection, is there someone who will monitor the data over time?
Has expiry

4) Using the **[category system of data types](https://wiki.mozilla.org/Firefox/Data_Collection)** on the Mozilla wiki, what collection type of data do the requested measurements fall under?
Type 2

5) Is the data collection request for default-on or default-off?
Default on

6) Does the instrumentation include the addition of **any *new* identifiers** (whether anonymous or otherwise; e.g., username, random IDs, etc.  See the appendix for more details)?
No

7) Is the data collection covered by the existing Firefox privacy notice?
Yes

8) Does there need to be a check-in in the future to determine whether to renew the data?
Will check-in at expiry

9) Does the data collection use a third-party collection tool?
No

**Pull request merged**